rfcs/0012-external-function.md

```diff
@@ -0,0 +1,163 @@
+---
+feature: external-function-framework
+authors:
+  - "TennyZhuang"
+start_date: "2022/11/04"
+---
+
+# External Function Framework
+
+## Summary
+
+We will introduce a gRPC-based external function framework in the kernel part. **The framework is private to users and may change later**.
+
+## Motivation
+
+The RFC is a preparation RFC for external UDFs with heavy runtime or complicated environments.
+
+The RFC will not answer the following questions:
+
+* How to define the UDFs?
+* How to execute the UDFs?
+
+## Design
+
+### External function protocol
+
+We will add a simple protocol for external function servers to implement:
+
+```protobuf
+message BatchExternalUdfRequest {
+    repeated bytes columns = 1;
```

Comment:
What's the internal encoding of input/output columns?

Reply:
If we want users to decode/encode a column, I think it's better to use some open format like Arrow or Avro. Otherwise we have to provide SDKs for all languages.

Reply:
Yes, we prefer to use arrow, but we’ll try to implement it directly in our array implementation.